### DISCUSSION TITLE
Looking for help getting started with bringing a .jar into a c# .net project

### DISCUSSION THREAD

**Discussion by tovversh** [ikvmnet/ikvm#350] [Q&A]
I've been working away at this for a couple of days, and I'm sure I'm missing something simple, but I just can't figure out what it is.  

So to start off I have a .jar file from an outside source that I would like to run with my c# .net project.  IKVM seems to be designed for that so I've tried to get it to work without much luck

So here is what I did:  ( Using Visual Studio 2019 )
1. Created a new Class Library ( .Net Framework )
	- Set the .Net Framework to 4.7.2  ( To match the project that will use the resulting library)
2. Added the NuGet package for IKVM (8.5.1 )
3. Added the Nuget package for IKVM.MSBuild (8.5.1 )
4. Saved and closed the solution.
5. Opened the .proj file in the editor and added:
  <ItemGroup>
    <IkvmReference Include="Reporter.jar">
      <AssemblyName>Reporter</AssemblyName>
      <Compile>Reporter.jar</Compile>
    </IkvmReference>
  </ItemGroup>
6. I opened the default Class1.cs and added a #using to the namespace from the .jar and added a function to the class to call the 'main' from the .jar.
7. Then I did a full rebuild.

The rebuild doesn't recognize the namespace from the .jar and fails.  The only error it gives me is not recognizing the namespace.  I see no indication that anything is happening with regards to the IkvmReference to bring in the contents of the .jar.

I've ran the .jar file with the java.exe included with IKVM and it operated successfully, so the .jar is basically good.  

At this point I'm not sure what to do next.  I'm feeling like I'm missing something simple, but haven't been able to figure it out.  Any help would be greatly appreciated.

**Comment by wasabii** [comment]
Make sure the project is a SDK style project. We do not support old style csproj formats.

**Reply by tovversh** [reply]
First let me thank you for your reply.  Any help is good help and this at least got me closer to where I'm trying to go.

I was able to build a dll from the .jar file using a SDK style project.  

I used these options for the item group:
```
	<ItemGroup>
		<IkvmReference Include="Reporter.jar">
			<AssemblyName>Reporter</AssemblyName>
			<AssemblyVersion>2.2.0.0</AssemblyVersion>
			<AssemblyFileVersion>2.2.0.0</AssemblyFileVersion>
			<Compile>Reporter.jar</Compile>
		</IkvmReference>
	</ItemGroup>
```

So that's good.  Maybe?  At least it gets me closer.

Now that I have the DLL, are there any other dependencies or requirements for using it in another class library or application?  Can I just include the DLL as a reference in my project, or do I need to include IKVM again?  Does the consuming project need to be a SDK style project?

**Reply by wasabii** [reply]
If you're using IkvmReference, as opposed to for instance, MavenReference, it functions the same as a plain.Net reference. It is not transitive. Every project should include IkvmReference.

IkvmReference is not designed to produce DLLs that you then copy out elsewhere.

The only thing you need to distribute is the Jar.

**Reply by tovversh** [reply]
Thank you again for your help.  Sorry to be slow getting back to this, it's kind of a side project I'm working on without a lot of time available.

I think I understand what you're saying here.  So I went back and created a simple stand alone project to test this.  Basically, I setup the following:
- c# wpf application
- I included the IKVM nuget project
- Added the IKVMReference to the project file
- Added the #using to the expected namespace
- called the .main() with a basic command ( in this case getVersion) if I run this command through the java command line I get the version back in the console.

The project compiled, however when it runs and hits the call to main(), I see nothing dumped to the debug window, and the application immediately closes.  I've made sure the .jar file is included in the project and is in the debug folder when I run the application.  

So I'm kind of back to not knowing how to proceed.  I need to be able to get the results that the .jar would usually dump to the console and I can't have it exiting my application when it finishes, as I want to process the data it returns.

Again, I really appreciate any help here.

**Reply by wasabii** [reply]
No problem.

At this point though I'm going to need an example project to check.

Unless you can get the debugger attached, make sure exceptions are hit, etc.

**Reply by tovversh** [reply]
No problem.  I ran it in Visual Studio in debug.  It hits the break point to call the .main and then exits when before the next line is hit.

Here's a zip of the project.  It is as bare bones as I can get it.  Many thanks for taking a look at it.

[TestJar.zip](https://github.com/ikvmnet/ikvm/files/11716282/TestJar.zip)

The call is Reporter.main in the MainWindow() of the MainWindow.xaml.cs

**Reply by wasabii** [reply]
So you made a Windows app, which doesn't show the console, that calls code that calls System.exit(0).

So it does exactly that.

**Reply by tovversh** [reply]
Ah, that would make sense.  I am not the source of the Jar and I don't have access to it's source ( though I will run around and check).  I just want a way to run it and get back the results without needing java installed.  I don't suppose there is a way to trap or block the System.exit(0) call?

**Reply by wasabii** [reply]
Nothing that would be special to IKVM.

But you don't have to call main.

**Reply by wasabii** [reply]
You can decompile it and take a look. Either in a Java tool, or the DLL in a .net tool

**Reply by tovversh** [reply]
Thanks a lot for the tips.  I was able to get into the jar file, decompile and edit the classes and remove a couple of the offending System.exit calls and got the resulting jar to run without forcing my application to exit.  So that should resolve that problem in principle.

That leaves me with one remaining issue.  It looks like the jar is directing many important bits of output through System.out.println.  I am not seeing them in the Visual Studio debug window when I run the application.  Do you know where IKVM would direct that output and if there would be any way in the c# that I might capture that.  

Again, thank you very much for your help.

**Reply by wasabii** [reply]
IKVM doesn't do anything special to it. It's Console.Output. stdout.

Which doesn't show in a Win Exe where there is no attached console.
